### PR TITLE
Triangles and meshes

### LINKED ISSUE #10: Add triangles and meshes

### PR DESCRIPTION
**Pull request by andylvua**
**This pull request adds support for triangles, meshes, and .obj files to the project.**

Closes #10

### CODE REVIEW

File: inc/objects/triangle.hpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#include "object.hpp"
+
+struct Vertex {
+public:
+    glm::vec3 position;
+    glm::vec3 normal;
+    glm::vec2 uv;
+};
+
+class Triangle : public Object {
```
Comment (codefloww) [inline inc/objects/triangle.hpp]: Consider remaking triangle not as object child

**Review by codefloww — APPROVED** [review]
Looks okay, but definitely need optimizations

File: inc/objects/triangle.hpp (hunk 1 above)
Comment (andylvua) [reply inc/objects/triangle.hpp]: I listen and obey, my review Lord - 54f56f3

**Comment by andylvua** [comment]
@codefloww Bon appetit, my review Lord!

File: src/objects/triangle.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    vertices_m[1] = v2;
+    vertices_m[2] = v3;
+
+    e1_m = vertices_m[1].position - vertices_m[0].position;
+    e2_m = vertices_m[2].position - vertices_m[0].position;
+    normal_m = glm::normalize(glm::cross(e1_m, e2_m));
+}
+
+bool Triangle::testIntersections(const Ray &cast_ray, glm::vec3 &int_point, glm::vec3 &loc_normal) const {
+    constexpr double kEpsilon = 0.0000001;
```
Comment (codefloww) [inline src/objects/triangle.cpp]: Guess you can keep epsilon consistent across all project, no need to have such precision
```suggestion
    constexpr double kEpsilon = 0.0001;
```

File: src/objects/trianglemesh.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                        glm::vec3{nx, ny, nz},
+                        glm::vec2{tx, ty}
+                );
+            }
+            offset += vertex;
+        }
+    }
+
+    for (int i = 0; i < vertices.size() / 3; ++i) {
+        triangles.push_back(std::make_shared<Triangle>(
```
Comment (codefloww) [inline src/objects/trianglemesh.cpp:62]:
```suggestion
        triangles.emplace_back(std::make_shared<Triangle>(
```

File: src/objects/trianglemesh.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                                     glm::vec3 &loc_color) const {
+    bool hit = false;
+    float closest_hit = std::numeric_limits<float>::max();
+    Ray local_ray = transformation_m.applyTransform(cast_ray, Direction::BACKWARD);
+
+    for (const auto &tri: triangles) {
+        glm::vec3 tri_int_point, tri_loc_normal;
+        if (tri->testIntersections(local_ray, tri_int_point, tri_loc_normal)) {
+            tri_int_point = transformation_m.applyTransform(tri_int_point, Direction::FORWARD);
+            tri_loc_normal = glm::normalize(transformation_m.applyTransform(tri_loc_normal, Direction::FORWARD));
```
Comment (codefloww) [inline src/objects/trianglemesh.cpp]: Can move out normalize into the if scope to reduce number of heave normalization calls

File: src/objects/trianglemesh.cpp (hunk 3 above)
Comment (andylvua) [reply src/objects/trianglemesh.cpp:62]: I would suspect it to be equivalent to a push_back, with the drawback of being less explicit. When you actually want to copy-construct from a temporary (what is happening here), I think the push family of methods is the canonical way of doing this

File: src/objects/trianglemesh.cpp (hunk 4 above)
Comment (andylvua) [reply src/objects/trianglemesh.cpp]: Fixed in 2bb0568

File: src/objects/triangle.cpp (hunk 2 above)
Comment (andylvua) [reply src/objects/triangle.cpp]: Updated in 1e821ca

File: src/objects/trianglemesh.cpp (hunk 3 above)
Comment (codefloww) [reply src/objects/trianglemesh.cpp:62]: As we won't use those shared pointers nowhere else, except that list, it looks that the emplacing them is a valid options as well. But anyway, it's up to you

**Review by codefloww — APPROVED** [review]
LGFM

**Pull request merged**